src/AutoRest.CSharp/Common/Generation/Types/TypeFactory.cs

```diff
@@ -200,19 +203,43 @@ public static CSharpType GetElementType(CSharpType type)
         internal static bool IsDictionary(CSharpType type)
             => IsReadOnlyDictionary(type) || IsReadWriteDictionary(type);
 
+        internal static bool IsDictionary(CSharpType type, [MaybeNullWhen(false)] out CSharpType keyType, [MaybeNullWhen(false)] out CSharpType valueType)
+        {
+            if (IsReadOnlyDictionary(type) || IsReadWriteDictionary(type))
+            {
+                keyType = type.Arguments[0];
+                valueType = type.Arguments[1];
+                return true;
+            }
+
+            keyType = null;
+            valueType = null;
+            return false;
+        }
+
         internal static bool IsReadOnlyDictionary(CSharpType type)
             => type.IsFrameworkType && type.FrameworkType == typeof(IReadOnlyDictionary<,>);
 
         internal static bool IsReadWriteDictionary(CSharpType type)
             => type.IsFrameworkType && (type.FrameworkType == typeof(IDictionary<,>) || type.FrameworkType == typeof(Dictionary<,>));
 
         internal static bool IsList(CSharpType type)
-            => IsReadOnlyList(type) || IsReadWriteList(type);
+            => IsList(type, out _);
+
+        internal static bool IsList(CSharpType type, [MaybeNullWhen(false)] out CSharpType elementType)
+        {
+            if (IsReadOnlyList(type) || IsReadWriteList(type))
```

Comment:
I think we can put check if it is List in `bool IsList(CSharpType type)` function, and call the funtion here.

Reply:
Fixed